spec/models/ext_management_system_spec.rb

```diff
@@ -396,6 +396,168 @@
+                               :zone_before_pause => zone,
+                               :zone              => Zone.maintenance_zone,
+                               :enabled           => false)
+
+      child = FactoryGirl.create(:ext_management_system,
+                                 :zone_before_pause => zone,
+                                 :zone              => Zone.maintenance_zone,
+                                 :enabled           => false)
+      ems.child_managers << child
+
+      ems.resume!
+
+      expect(ems.enabled).to eq(true)
+      expect(ems.zone).to eq(zone)
+      expect(ems.zone_before_pause).to be_nil
+
+      child.reload
+      expect(child.enabled).to eq(true)
+      expect(child.zone).to eq(zone)
+      expect(child.zone_before_pause).to be_nil
+    end
+
+    it "doesn't change zone when ems was disabled before" do
+      zone = FactoryGirl.create(:zone)
+      ems = FactoryGirl.create(:ext_management_system,
+                               :zone_before_pause => nil,
+                               :zone              => zone,
+                               :enabled           => false)
+
+      ems.resume!
+
+      expect(ems.enabled).to eq(true)
+      expect(ems.zone).to eq(zone)
+    end
+
+    it 'enables an ems and moves child managers created by ensure_managers() to original zone' do
+      zone  = FactoryGirl.create(:zone)
+      # Cannot create paused child managers this way, so we first pause! (tested yet) and then resume!
+      emses = {
+        :amazon_cloud    => FactoryGirl.create(:ems_amazon,          :zone => zone),
+        :azure_cloud     => FactoryGirl.create(:ems_azure,           :zone => zone),
+        :google_cloud    => FactoryGirl.create(:ems_google,          :zone => zone),
+        :openstack_cloud => FactoryGirl.create(:ems_openstack,       :zone => zone),
+        :openstack_infra => FactoryGirl.create(:ems_openstack_infra, :zone => zone),
+        :vmware_cloud    => FactoryGirl.create(:ems_vmware_cloud,    :zone => zone),
+      }
+
+      # managers with child managers created with ensure_managers() callback has own callback to change zone
+      # so it should be tested separately
+      emses.each_value do |manager|
+        manager.pause!
+        manager.reload
+
+        manager.resume!
+        manager.reload
+
+        expect(manager.network_manager.zone_before_pause).to be_nil
+        expect(manager.network_manager.zone).to eq(zone)
+      end
+
+      # The same for storage managers, i.e. amazon
+      %i(amazon_cloud openstack_cloud).each do |manager_type|
+        emses[manager_type].storage_managers.each do |storage_manager|
+          expect(storage_manager.zone_before_pause).to be_nil
+          expect(storage_manager.zone).to eq(zone)
+        end
+      end
+    end
+  end
+
+  context "changing zone" do
+    before { Zone.seed }
+
+    it 'is allowed when enabled' do
+      zones = FactoryGirl.create_list(:zone, 2)
```

Comment:
Do you need 2 zones here?  Since `Zone.seed` is in the `before` block, you have the default zone that could be used.

Reply:
It's about checking 2 random normal (== non-maintenance) zones. `Zone.seed` is there because of correct validation conditions in EMS.